assets/src/stories-editor/blocks/amp-story-page-attachment/attachment-content.js

```diff
@@ -0,0 +1,180 @@
+/**
+ * External dependencies
+ */
+import PropTypes from 'prop-types';
+
+/**
+ * Internal dependencies
+ */
+import { PostSelector } from '../../components';
+import useElementClickDetector from './element-click-detector';
+
+/**
+ * WordPress dependencies
+ */
+import { __ } from '@wordpress/i18n';
+import apiFetch from '@wordpress/api-fetch';
+import { RichText } from '@wordpress/block-editor';
+import { Button } from '@wordpress/components';
+import { RawHTML, useEffect, useRef, useState } from '@wordpress/element';
+import { ENTER, SPACE } from '@wordpress/keycodes';
+
+const AttachmentContent = ( props ) => {
+	const [ selectedPost, setSelectedPost ] = useState( null );
+	const [ failedToFetch, setFailedToFetch ] = useState( false );
+	const [ searchValue, setSearchValue ] = useState( '' );
+	let fetchRequest, isStillMounted;
+
+	const {
+		attributes,
+		setAttributes,
+		toggleAttachment,
+	} = props;
+
+	const {
+		attachmentClass,
+		postId,
+		postType,
+		title,
+		wrapperStyle,
+	} = attributes;
+
+	const fetchSelectedPost = () => {
+		isStillMounted = true;
```

Comment:
Looks like isMounted checks are considered an [anti-pattern](https://reactjs.org/blog/2015/12/16/ismounted-antipattern.html).

A possible alternative using a custom hook: https://github.com/rajeshnaroth/react-cancelable-promise-hook

I also stumbled upon https://github.com/dai-shi/react-hooks-async, but that seems a bit overkill for this use case.

Reply:
>Looks like isMounted checks are considered an anti-pattern.

It looks to be that's specifically for `isMounted()` and it actually mentions below to use your own check for if the component is still mounted:

> An easy migration strategy for anyone upgrading their code to avoid isMounted() is to track the mounted status yourself. Just set a _isMounted property to true in componentDidMount and set it to false in componentWillUnmount, and use this variable to check your component’s status.

In this component's case we do need to ensure that the component is still mounted since the fetch request might take longer time and the component might already get unmounted meanwhile, it's not necessarily a faulty logic within the component that creates setting state when the component is already unmounted (as mentioned in the post). Thoughts?

Reply:
In this instance this pattern remembering "mounted" state is definitely acceptable, as it is the only way to know whether the callback still needs to be run. 

The `react-cancelable-promise-hook` does the exact same thing internally. The module however does not work.

Using this module, you cannot tell the difference between the promise cancelling because the original promise failed (bad fetch) or because it has been cancelled. Not easily anyway.

They do attempt to [pass an object as the error message](https://github.com/rajeshnaroth/react-cancelable-promise-hook/blob/master/index.js#L8-L9) to be able to distinguish based off this property, but that fails as the JS engine (at least Chrome) internally converts the message to a string:

```
assert(new Error({foo: 1}).message === '[object Object]')
```

Creating a helper is definitely a good idea, but it has to be done correctly. Creating a custom error object to throw adding an `isCancelled` attribute. If you even need to throw an error and not just silently ignore the cancelled promise return.

I see different ways of creating such a helper, but thinking deeper about it, it's actually quite a bit tricky. What's the proper API for such a hook? Let's wait until we see a few more use-cases for such behaviour before trying to refactor.

I'd just go for the simple property that you have now, and worry about refactoring later. I don't event think it's worthy of a tech debt ticket.

Reply:
> I'd just go for the simple property that you have now, and worry about refactoring later. I don't event think it's worthy of a tech debt ticket.

Works for me. Thanks to both of you!